src/smtencoding/FStar.SMTEncoding.Term.fst

```diff
@@ -638,8 +636,36 @@ let constructor_to_decl rng constr =
                     Some "Discriminator definition") in
         def in
     let projs = injective_constructor rng (constr.constr_name, constr.constr_fields, sort) in
+    let base =
+      if not constr.constr_base
+      then []
+      else (
+        let arg_sorts =
+          constr.constr_fields 
+          |> List.filter (fun f -> f.field_projectible)
+          |> List.map (fun _ -> Term_sort)
+        in
+        let base_name = constr.constr_name ^ "@base" in
+        let decl = DeclFun(base_name, arg_sorts, Term_sort, Some "Constructor base") in
+        let formals = List.mapi (fun i _ -> mk_fv ("x" ^ string_of_int i, Term_sort)) constr.constr_fields in
+        let constructed_term = mkApp(constr.constr_name, List.map (fun fv -> mkFreeV fv norng) formals) norng in
+        let inj_formals = List.flatten <| List.map2 (fun f fld -> if fld.field_projectible then [f] else []) formals constr.constr_fields in
+        let base_term = mkApp(base_name, List.map (fun fv -> mkFreeV fv norng) inj_formals) norng in
+        let eq = mkEq(constructed_term, base_term) norng in
+        let guard = mkApp(discriminator_name constr, [constructed_term]) norng in
+        let q = mkForall rng ([[constructed_term]], formals, mkImp (guard, eq) norng) in
+        //forall (x0...xn:Term). {:pattern (C x0 ...xn)} is-C (C x0..xn) ==> C x0..xn == C-base x2 x3..xn
```

Comment:
When constr.constr_base is set, generate an axiom to state that all the non-projectible fields are irrelevant